libs/testproxy/server.go

```diff
@@ -0,0 +1,177 @@
+package testproxy
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/databricks/cli/internal/testutil"
+	"github.com/databricks/cli/libs/testserver"
+	"github.com/databricks/databricks-sdk-go/apierr"
+	"github.com/databricks/databricks-sdk-go/client"
+	"github.com/databricks/databricks-sdk-go/config"
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type ProxyServer struct {
+	*httptest.Server
+
+	t      testutil.TestingT
+	router *mux.Router
+
+	apiClient        *client.DatabricksClient
+	RequestCallback  func(request *testserver.Request)
+	ResponseCallback func(request *testserver.Request, response *testserver.EncodedResponse)
+}
+
+// This creates a reverse proxy server that sits in front of a real Databricks
+// workspace. This is useful for recording API requests and responses in
+// integration tests.
+//
+// Note: We cannot simply proxy the request from a localhost URL to a real
+// workspace. This is because auth resolution in the Databricks SDK relies
+// what the URL actually looks like to determine the auth method to use.
+// For example, in OAuth flows, the SDK can make requests to different Microsoft
+// OAuth endpoints based on the nature of the URL.
+// For reference, see:
+// https://github.com/databricks/databricks-sdk-go/blob/79e4b3a6e9b0b7dcb1af9ad4025deb447b01d933/common/environment/environments.go#L57
+func New(t testutil.TestingT) *ProxyServer {
+	router := mux.NewRouter()
+	server := httptest.NewServer(router)
+	t.Cleanup(server.Close)
+
+	s := &ProxyServer{
+		Server: server,
+		t:      t,
+		router: router,
+	}
+
+	// Create an API client using the current authentication context.
+	// In CI test environments this would read the appropriate environment
+	// variables.
+	var err error
+	s.apiClient, err = client.New(&config.Config{})
+	require.NoError(t, err)
+
+	// Set up the proxy handler as the default handler for all requests.
+	router.NotFoundHandler = http.HandlerFunc(s.proxyToCloud)
+	return s
+}
+
+func (s *ProxyServer) reqBody(r testserver.Request) any {
+	// The SDK expects the query parameters to be specified in the "request body"
+	// argument for GET, DELETE, and HEAD requests in the .Do method.
+	if r.Method == "GET" || r.Method == "DELETE" || r.Method == "HEAD" {
+		queryParams := make(map[string]any)
+		for k, v := range r.URL.Query() {
+			queryParams[k] = v[0]
+		}
+		return queryParams
+	}
+
+	// The SDK does not support directly passing a JSON serialized request
+	// body. So we convert it to a map if the body is a JSON object.
+	if json.Valid(r.Body) {
```

Comment:
Why don't we pass raw body in all cases? unmarshalling / marshalling without changes is highly undesirable.

Reply:
Like the comment mentions, we cannot. SDK does not work with a raw body that contains JSON bytes.

Reply:
That's very surprising. Is this documented behaviour? For my knowledge, could you point to docs/source?

There is another client in SDK -- I'm using it here https://github.com/databricks/cli/pull/2791/files#diff-de02fc473c0f577a6412715ed686456e0cc18ad0e9a0a594ec4ad50dccadd66aR27 and it does not have this behaviour

Reply:
I think bytes don't work because we only set the application type header in the default case here: 
[](https://github.com/databricks/databricks-sdk-go/blob/4f0fe8782d4f09bf3929b00b8fedb70895bc30cb/common/request.go#L50)

I don't recall which case was failing because of this, and existing tests pass without it.

Reply:
I'm fine with removing it; we can add it back in once we run into the error case again; let me know which one you prefer. I'd prefer to keep it as is.

Reply:
Alternately we could set the content type header, taking a look whether that works.

Reply:
Removed this. The pipeline crud test case provides coverage for JSON bodies so it should be fine.